servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -359,12 +483,32 @@ void markInactive() {
             }
         }
 
+        @SuppressWarnings("PMD.CollapsibleIfStatements")
+        void markExpired() {
+            stateUpdater.set(this, connections.length == 0 ? State.CLOSED : State.EXPIRED);
+            if (state == State.CLOSED) {
+                // if in the meantime a connection was added, we shall close it gracefully
+                this.closeAsyncGracefully().subscribe();
+            }
+        }
+
         boolean isInactive() {
-            return connections == CLOSED_ARRAY;
+            return state == State.CLOSED;
+        }
+
+        boolean isActive() {
+            return state == State.ACTIVE;
+        }
+
+        boolean isExpired() {
+            return state == State.EXPIRED;
         }
 
         boolean addConnection(C connection) {
             for (;;) {
+                if (state == State.CLOSED) {
```

Comment:
if the state is closed won't this host be deferenced (eventually)? if so do we need this check here on every `addConnection`? at best this is only a best effort check anyways due to independent volatile variables IIUC.

Reply:
That is true, I added a new test for concurrent behaviour that validates removing this best effort check does not cause unexpected behaviour.